networkx/algorithms/approximation/steinertree.py

```diff
@@ -113,10 +113,24 @@ def _kou_steiner_tree(G, terminal_nodes, weight):
 
 
 def _remove_nonterminal_leaves(G, terminals):
-    terminals_set = set(terminals)
-    for n in list(G.nodes):
-        if n not in terminals_set and G.degree(n) == 1:
-            G.remove_node(n)
+    terminal_set = set(terminals)
+    leaves = {
+        n
+        for n in G.nodes()
+        if len(list(G.neighbors(n))) - list(G.neighbors(n)).count(n) == 1
+    }
```

Comment:
I think the following might be a bit easier to read:

```suggestion
    leaves = {n for n in G if len(set(G[n]) - {n}) == 1}
```